tools/yulPhaser/Population.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Population.h>
+
+#include <tools/yulPhaser/Program.h>
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::phaser
```

Comment:
```suggestion
using namespace solidity::phaser;
```

Reply:
Right. I didn't notice you were using this style. I'll switch to it in my code.

Reply:
Done. There's one downside to this style though. It forces me to add forward declarations of all `friend`s which usually requires forward declaring the class too.

```
struct Individual;
class Population;
std::ostream& operator<<(std::ostream& _stream, Individual const& _individual);
std::ostream& operator<<(std::ostream& _stream, Population const& _population);
```

I don't need to do this if the definition is wrapped in a namespace too.

I wondered why this did not came up in the rest of the codebase. Looks like you mostly just define `operator<<` inline. I didn't want to pollute the header with its implementation details though.

Reply:
Hm, I don't see why a change in the cpp file demands a change in the header. This probably has not come up because we don't really use `operator<<`.

Reply:
I just checked and the forward declaration does not really need to be in the header. Looks like I could have put it in the `.cpp` file just as well. The header was a natural choice for a declaration but I just realized that putting it in the `.cpp` file would have saved me the trouble of having to forward-declare the class. I'll move the declarations there. Still, a declaration inside the namespace is required.

It has not come up so far because the existing code always defines friends inline inside the namespace (inside or outside of class body but still in the header file). And a definition outside of the namespace requires a declaration in that namespace even if you do `using namespace`:

``` c++
namespace NX { void foo(); }
namespace NY {}

using namespace NX;
using namespace NY;

void NX::foo() {}
void NY::bar() {} // ERROR: 'void NY::bar()' should have been declared inside 'NY'
```